p2pserver/net/netserver/nbr_peers.go

```diff
@@ -0,0 +1,204 @@
+/*
+ * Copyright (C) 2018 The ontology Authors
+ * This file is part of The ontology library.
+ *
+ * The ontology is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * The ontology is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with The ontology.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package netserver
+
+import (
+	"net"
+	"sync"
+	"sync/atomic"
+
+	"github.com/ontio/ontology/p2pserver/peer"
+
+	comm "github.com/ontio/ontology/common"
+	"github.com/ontio/ontology/common/log"
+	"github.com/ontio/ontology/p2pserver/common"
+	"github.com/ontio/ontology/p2pserver/message/types"
+)
+
+// Conn is a net.Conn wrapper to do some clean up when Close.
+type Conn struct {
+	net.Conn
+	session   uint64
+	id        common.PeerId
+	closed    bool
+	netServer *NetServer
+}
+
+// Close overwrite net.Conn
+func (self *Conn) Close() error {
```

Comment:
When is this Conn closed?

Reply:
this wrapped Conn is stored in `Link` using RAII pattern, when Link has error, it will close the underline socket and clean the associated resources in nbr_peers and connection controller together.
the original code has double close issues.